runtime/common/src/zenlink.rs

```diff
@@ -0,0 +1,257 @@
+pub fn zenlink_id_to_currency_id(asset_id: ZenlinkAssetId) -> Result<CurrencyId, ()> {
+	let _index = asset_id.asset_index;
+	let disc = ((_index & 0x0000_0000_0000_ff00) >> 8) as u8;
+	let symbol = (_index & 0x0000_0000_0000_00ff) as u8;
+	match disc {
+		0 => Ok(CurrencyId::Native),
+		1 => Ok(CurrencyId::XCM(symbol)),
+		2 => match symbol {
+			0 => Ok(CurrencyId::Stellar(Asset::StellarNative)),
+			1 => Ok(CurrencyId::Stellar(Asset::AlphaNum4 { code: *b"USDC", issuer: USDC_ISSUER })),
+			2 => Ok(CurrencyId::Stellar(Asset::AlphaNum4 { code: *b"TZS\0", issuer: TZS_ISSUER })),
+			3 => Ok(CurrencyId::Stellar(Asset::AlphaNum4 { code: *b"BRL\0", issuer: BRL_ISSUER })),
+			_ => return Err(()),
+		},
+		6 => {
+			let token1_id = ((_index & 0x0000_0000_00FF_0000) >> 16) as u8;
+			let token1_type = ((_index & 0x0000_0000_FF00_0000) >> 24) as u8;
+
+			let token2_id = ((_index & 0x0000_00FF_0000_0000) >> 32) as u8;
+			let token2_type = ((_index & 0x0000_FF00_0000_0000) >> 40) as u8;
+
+			Ok(CurrencyId::ZenlinkLPToken(token1_id, token1_type, token2_id, token2_type))
+		},
+		_ => Err(()),
+	}
+}
+
+pub fn currency_id_to_zenlink_id(
+	currency_id: CurrencyId,
+	parachain_id: u32,
+) -> Result<ZenlinkAssetId, ()> {
+	let disc = discriminant(&currency_id) as u64;
+	match currency_id {
+		CurrencyId::Native =>
+			Ok(ZenlinkAssetId { chain_id: parachain_id, asset_type: NATIVE, asset_index: 0 as u64 }),
+		CurrencyId::XCM(token_id) => Ok(ZenlinkAssetId {
+			chain_id: parachain_id,
+			asset_type: LOCAL,
+			asset_index: ((disc << 8) + token_id as u64) as u64,
+		}),
+		CurrencyId::Stellar(asset) => {
+			let _id = match asset {
+				Asset::StellarNative => 0u64,
+				Asset::AlphaNum4 { code, .. } => match &code {
+					b"USDC" => 1u64,
+					b"TZS\0" => 2u64,
+					b"BRL\0" => 3u64,
+					_ => return Err(()),
+				},
+				_ => return Err(()),
+			};
+			Ok(ZenlinkAssetId {
+				chain_id: parachain_id,
+				asset_type: LOCAL,
+				asset_index: ((disc << 8) + _id) as u64,
+			})
+		},
+		CurrencyId::ZenlinkLPToken(token1_id, token1_type, token2_id, token2_type) => {
+			let _index = ((disc as u64) << 8) +
+				((token1_id as u64) << 16) +
+				((token1_type as u64) << 24) +
+				((token2_id as u64) << 32) +
+				((token2_type as u64) << 40);
+			Ok(ZenlinkAssetId { chain_id: parachain_id, asset_type: LOCAL, asset_index: _index })
+		},
+	}
+}
+
+#[cfg(test)]
+mod zenlink_tests {
+	use super::*;
+
+	#[test]
+	fn convert_zenlink_native_to_native_currency() {
+		// Native ZenlinkAsset index = 0x0000_0000_0000_0000
+		let _index = 0 as u64;
+		let fake_native_asset =
+			ZenlinkAssetId { chain_id: 1000u32, asset_type: NATIVE, asset_index: 0 as u64 };
+		let currency: Result<CurrencyId, _> = zenlink_id_to_currency_id(fake_native_asset);
+		assert_eq!(currency, Ok(CurrencyId::Native));
+	}
+
+	#[test]
+	fn convert_zenlink_xcm_to_xcm_currency() {
+		// XCM(0) ZenlinkAsset index = 0x0000_0000_0000_0100
+		let _index = 0x0100 as u64;
+		let fake_native_asset =
```

Comment:
Nit: You probably re-used the name `fake_native_asset` when copy-pasting code for convenience. It doesn't make sense in all functions though.  I'd maybe rename it to `fake_zenlink_asset` in all tests so that it's 'correct' and you don't have to spend time on writing a proper explicit name haha.

Reply:
That's true, my bad haha :joy:.